### PR TITLE
Feature: Connection Time Out feature update

### LINKED ISSUE #312: Ability to add connection timeout
**Tell us about the task you want to perform and are unable to do so because the feature is not available**
User should be provided with an option in Settings to add connection timeout to terminate the request if time elapsed is more than this timeout.

### PR DESCRIPTION
**Pull request by nikkivirtuoso**
## PR Description

this PR is not complete creted for a review

## Related Issues

- Related Issue #312
- Closes #312

### Checklist
- [x] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [x] I have run the tests (`flutter test`) and all tests are passing

## Added/updated tests?
_We encourage you to add relevant test cases._

- [ ] Yes
- [x] No, and this is why: its a feature

### CODE REVIEW

**Comment by nikkivirtuoso** [comment]
@animator sir can you please review this as it is showing the time out error but the thing is once the response is recieved from thr API side the timeout error is replaced by the actual response please guide me to tackle this problem i have used flag but the code is changing the its value

**Comment by nikkivirtuoso** [comment]
@animator sir the feature is successfully implemented please have a look

**Comment by animator** [comment]
@nikkivirtuoso Please cleanup the code. The quality of code you currently submitted is not acceptable. Do some research on how you can write clean codes.

**Comment by nikkivirtuoso** [comment]
sorry sir @animator i forgot to remove the commented code and did some spelling mistakes now i have removed the unnecessary things please have a look

**Comment by nikkivirtuoso** [comment]
@animator @ashitaprasad please have a look i have cleaned the code

**Comment by nikkivirtuoso** [comment]
@animator i have stored the value of connectiontimeout in a timeout variable if still my code is unacceptable please guide me so that I can get to know which part i need to make changes

File: lib/models/settings_model.dart [hunk 1]
```diff
@@ -54,6 +58,7 @@ class SettingsModel {
     final height = data["height"] as double?;
     final dx = data["dx"] as double?;
     final dy = data["dy"] as double?;
+    final connectionTimeout = data["connectionTimeout"] as int? ?? 10;
```
Comment (animator) [inline lib/models/settings_model.dart:61]: ☝️

File: lib/providers/collection_providers.dart [hunk 2]
```diff
@@ -177,14 +177,14 @@ class CollectionStateNotifier
     );
 
     RequestModel requestModel = state![id]!;
-
-    // set current model's isWorking to true and update state
+    SettingsModel settingsModel = ref.read(settingsProvider);
```
Comment (animator) [inline lib/providers/collection_providers.dart:180]: ☝️

File: lib/providers/collection_providers.dart [hunk 3]
```diff
@@ -177,14 +177,14 @@ class CollectionStateNotifier
...
     var map = {...state!};
     map[id] = requestModel.copyWith(isWorking: true);
     state = map;
 
     (http.Response?, Duration?, String?)? responseRec = await request(
       requestModel,
+      settingsModel,
```
Comment (animator) [inline lib/providers/collection_providers.dart:187]: ☝️

File: lib/providers/settings_providers.dart [hunk 4]
```diff
@@ -26,6 +26,7 @@ class ThemeStateNotifier extends StateNotifier<SettingsModel> {
     CodegenLanguage? defaultCodeGenLang,
     bool? saveResponses,
     bool? promptBeforeClosing,
+    int? connectionTimeout = 10,
```
Comment (animator) [inline lib/providers/settings_providers.dart:29]: ☝️

File: lib/providers/settings_providers.dart [hunk 5]
```diff
@@ -36,6 +37,7 @@ class ThemeStateNotifier extends StateNotifier<SettingsModel> {
       defaultCodeGenLang: defaultCodeGenLang,
       saveResponses: saveResponses,
       promptBeforeClosing: promptBeforeClosing,
+      connectionTimeout: connectionTimeout ?? state.connectionTimeout,
```
Comment (animator) [inline lib/providers/settings_providers.dart:40]: ☝️

File: lib/services/http_service.dart [hunk 6]
```diff
@@ -56,8 +59,10 @@ Future<(http.Response?, Duration?, String?)> request(
               );
             }
           }
-          http.StreamedResponse multiPartResponse =
-              await multiPartRequest.send();
+          http.StreamedResponse multiPartResponse = await multiPartRequest
+              .send()
+              .timeout(Duration(
```
Comment (animator) [inline lib/services/http_service.dart:64]: ☝️

File: lib/services/http_service.dart [hunk 7]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
...
+              .patch(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
+                  milliseconds: timeOut));
           break;
         case HTTPVerb.delete:
-          response =
-              await http.delete(requestUrl, headers: headers, body: body);
+          response = await http
+              .delete(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
```
Comment (animator) [inline lib/services/http_service.dart:102]: ☝️

File: lib/services/http_service.dart [hunk 8]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
...
+          response = await http
+              .put(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
+                  milliseconds: timeOut));
           break;
         case HTTPVerb.patch:
-          response = await http.patch(requestUrl, headers: headers, body: body);
+          response = await http
+              .patch(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
```
Comment (animator) [inline lib/services/http_service.dart:96]: ☝️

File: lib/services/http_service.dart [hunk 9]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
...
+          response = await http
+              .post(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
+                  milliseconds: timeOut));
           break;
         case HTTPVerb.put:
-          response = await http.put(requestUrl, headers: headers, body: body);
+          response = await http
+              .put(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
```
Comment (animator) [inline lib/services/http_service.dart:90]: ☝️

File: lib/services/http_service.dart [hunk 10]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
...
         case HTTPVerb.head:
-          response = await http.head(requestUrl, headers: headers);
+          response = await http.head(requestUrl, headers: headers).timeout(
+              Duration(milliseconds: timeOut));
           break;
         case HTTPVerb.post:
-          response = await http.post(requestUrl, headers: headers, body: body);
+          response = await http
+              .post(requestUrl, headers: headers, body: body)
+              .timeout(Duration(
```
Comment (animator) [inline lib/services/http_service.dart:84]: ☝️

File: lib/services/http_service.dart [hunk 11]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
...
       switch (requestModel.method) {
         case HTTPVerb.get:
-          response = await http.get(requestUrl, headers: headers);
+          response = await http.get(requestUrl, headers: headers).timeout(
+              Duration(milliseconds: timeOut));
           break;
         case HTTPVerb.head:
-          response = await http.head(requestUrl, headers: headers);
+          response = await http.head(requestUrl, headers: headers).timeout(
+              Duration(milliseconds: timeOut));
```
Comment (animator) [inline lib/services/http_service.dart:79]: ☝️

File: lib/services/http_service.dart [hunk 12]
```diff
@@ -66,27 +71,42 @@ Future<(http.Response?, Duration?, String?)> request(
       }
       switch (requestModel.method) {
         case HTTPVerb.get:
-          response = await http.get(requestUrl, headers: headers);
+          response = await http.get(requestUrl, headers: headers).timeout(
+              Duration(milliseconds: timeOut));
```
Comment (animator) [inline lib/services/http_service.dart:75]: ☝️

File: lib/services/http_service.dart [hunk 13]
```diff
@@ -24,6 +26,7 @@ Future<(http.Response?, Duration?, String?)> request(
       Stopwatch stopwatch = Stopwatch()..start();
       var isMultiPartRequest =
           requestModel.requestBodyContentType == ContentType.formdata;
+          final timeOut = settingsModel.connectionTimeout ?? 30;
```
Comment (animator) [inline lib/services/http_service.dart:29]: ☝️

**Comment by animator** [comment]
@nikkivirtuoso I have added pointers to places where you need to think more on your approach.

**Comment by nikkivirtuoso** [comment]
@animator , so you are saying that i don't have to make changes in the existing file, i have to make another file and do all the stuff right ? something like wrapping.

**Comment by animator** [comment]
No. I am saying the changes pointed above are incorrect.

File: lib/providers/collection_providers.dart (hunk 2 above)
Comment (nikkivirtuoso) [reply lib/providers/collection_providers.dart:180]: @animator it is taking the user's data, what is wrong with this ?

**Pull request closed without merging**